core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -681,6 +694,17 @@ class GroupCoordinator(val brokerId: Int,
   def onCompleteJoin(group: GroupMetadata) {
     var delayedStore: Option[DelayedStore] = None
     group synchronized {
+      if (group.generationId == 0) {
+        if (group.newMemberAdded && time.milliseconds() < group.initialRebalanceTimeout) {
+          group.newMemberAdded = false
+          val delay = scala.math.min(groupConfig.groupInitialRebalanceDelayMs, group.initialRebalanceTimeout - time.milliseconds()).toInt
+          val groupKey = GroupKey(group.groupId)
+          joinPurgatory.watchOperation(new DelayedJoin(this, group, delay, group.initialRebalanceTimeout), Seq(groupKey))
```

Comment:
@guozhangwang - one issue with this is that the total delay is going to be `min(numMembers * configured-value, rebalanceTimeout)`
What we are trying to do is a little different, i.e, as each new member joins it should reset the delay to `min(configured-value, remainingTime)`

I'm also not convinced that this will work. On the empty group we add a `FurtherDelayedJoin` to the purgatory that is scheduled to run in `delayMs`. In the meantime another member gets added to the group. As the group is not empty we'd do a `tryCompleteElseWatch(DelayedJoin)` which would complete. So we don't get the desired outcome.

We could change the final part of  `prepareRebalance` to
```
if(previousState.eq(Empty)) 
  joinPurgatory.tryCompleteElseWatch(new FurtherDelayedJoin(..),..)
else if (group.generationId != 0)
  joinPurgatory.tryCompleteElseWatch(new DelayedJoin(..), ...)
```

But i think we'd also need to track somewhere when the last member joined the group as we only want to extend the delay from that point of time. So when the `FurtherDelayedJoin#onComplete` is triggered i thin the new delay should be calculated as
`min(configured-delay - (now - lastMemberJoinedTime), remainingMs)`